File: skills/captain-obvious/scripts/co_ts/coverage.mjs
import fs from 'node:fs';
import path from 'node:path';

export function loadCoverage(file, projectDir) {
  let raw;
  try { raw = fs.readFileSync(file, 'utf8'); } catch { return null; }
  const map = new Map();
  const norm = (p) => path.relative(projectDir, path.isAbsolute(p) ? p : path.resolve(projectDir, p))
    .split(path.sep).join('/');
  const put = (p, line, hits) => {
    const rel = norm(p);
    let m = map.get(rel);
    if (!m) { m = new Map(); map.set(rel, m); }
    m.set(line, Math.max(m.get(line) ?? 0, hits));
  };
  const trimmed = raw.trimStart();
  if (trimmed.startsWith('{') || trimmed.startsWith('[')) {
    let json;
    try { json = JSON.parse(raw); } catch { return null; }
    if (json && json.files && typeof json.files === 'object') {
      for (const [p, d] of Object.entries(json.files)) {
        for (const ln of d.executed_lines ?? []) put(p, ln, 1);
        for (const ln of d.missing_lines ?? []) put(p, ln, 0);
      }
    } else if (json && typeof json === 'object') {
      for (const [p, d] of Object.entries(json)) {
        if (!d || !d.statementMap || !d.s) continue;
        for (const [id, loc] of Object.entries(d.statementMap)) {
          const line = loc?.start?.line;
          if (line == null) continue;
          put(d.path ?? p, line, d.s[id] ?? 0);
        }
      }
    }
  } else {
    let cur = null;
    for (const line of raw.split(/\r?\n/)) {
      if (line.startsWith('SF:')) cur = line.slice(3).trim();
      else if (line.startsWith('DA:') && cur) {
        const [ln, hits] = line.slice(3).split(',');
        put(cur, parseInt(ln, 10), parseInt(hits, 10) || 0);
      } else if (line.startsWith('end_of_record')) cur = null;
    }
  }
  if (map.size === 0) return null;
  return {
    hits(relFile, line) {
      const m = map.get(String(relFile).split(path.sep).join('/'));
      if (!m) return undefined;
      return m.has(line) ? m.get(line) : undefined;
    },
  };
}

File: skills/captain-obvious/scripts/co_py/coverage.py
from __future__ import annotations
import json
import os

def load_coverage(path: str, root: str):
    """Load line coverage (the dynamic signal ICSE'19 uses to separate rotten
    from good). Supports coverage.py json, lcov (DA: records), and istanbul
    coverage-final.json. Returns {(relpath, line): hits} or None."""
    try:
        raw = open(path, encoding="utf-8").read()
    except OSError:
        return None
    cov: dict[tuple[str, int], int] = {}

    def put(absf: str, line: int, hits: int):
        rel = os.path.relpath(absf, root).replace(os.sep, "/")
        key = (rel, line)
        cov[key] = max(cov.get(key, 0), hits)

    def absol(p: str) -> str:
        return p if os.path.isabs(p) else os.path.join(root, p)

    stripped = raw.lstrip()
    if stripped[:1] in ("{", "["):
        try:
            data = json.loads(raw)
        except (ValueError, json.JSONDecodeError):
            return None
        if isinstance(data, dict) and isinstance(data.get("files"), dict):   # coverage.py json
            for p, d in data["files"].items():
                for ln in d.get("executed_lines", []):
                    put(absol(p), ln, 1)
                for ln in d.get("missing_lines", []):
                    put(absol(p), ln, 0)
        elif isinstance(data, dict):                                          # istanbul json
            for p, d in data.items():
                if not isinstance(d, dict) or "statementMap" not in d or "s" not in d:
                    continue
                for sid, loc in d["statementMap"].items():
                    ln = (loc or {}).get("start", {}).get("line")
                    if ln is None:
                        continue
                    put(absol(d.get("path", p)), ln, d["s"].get(sid, 0))
    else:                                                                     # lcov
        cur = None
        for line in raw.splitlines():
            if line.startswith("SF:"):
                cur = absol(line[3:].strip())
            elif line.startswith("DA:") and cur:
                parts = line[3:].split(",")
                try:
                    put(cur, int(parts[0]), int(parts[1]))
                except (ValueError, IndexError):
                    continue
            elif line.startswith("end_of_record"):
                cur = None
    return cov or None
